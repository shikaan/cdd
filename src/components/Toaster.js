import React from "react";
import {createPortal} from "react-dom";
import PropTypes from "prop-types";
import styled from "styled-components";

import {Size} from "~theme";

const toasterMaxWidth = new Size(38);

const Wrapper = styled.div(({theme, visible}) => `
  position: fixed;
  bottom: 6vh;
  width: 100%;
  transition: all 400ms;
  ${handleVisibility(theme, visible)};
`);

const handleVisibility = (theme, visible) => {
  const visibleStyle = `
    opacity: 1;
    transform: translateY(0);
    z-index: ${theme.zIndexRank.top};
  `;

  const invisibleStyle = `
    opacity: 0;
    transform: translateY(16px);
    z-index: ${theme.zIndexRank.bottom};
  `;

  return visible ? visibleStyle : invisibleStyle;
};

const Toaster = React.memo(styled.div(({theme}) => `
  background: ${theme.color.darkGrey};
  color: ${theme.color.white};
  border-radius: 2px;
  text-align: center;
  font-family: ${theme.typography.secondaryFont.fontFamily};
  font-size: ${theme.typography.baseFontSize};
  font-weight: 700;
  letter-spacing: 0.15px;
  line-height: ${new Size(4)};
  max-width: ${toasterMaxWidth};
  margin: auto;
`));

const ToasterPortal = ({children, visible}) => {
  return createPortal(
    <Wrapper visible={visible}>
      <Toaster>{children}</Toaster>
    </Wrapper>,
    document.body
  );
};

ToasterPortal.propTypes = {
  visible: PropTypes.bool
};

export default ToasterPortal;
