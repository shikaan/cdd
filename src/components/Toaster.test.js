import React from 'react'
import {shallow, mount} from "enzyme";
import Toaster from "./Toaster";
import 'jest-styled-components';

import themeMock from '~theme'


describe('Toaster', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Toaster/>)

    expect(wrapper).toMatchSnapshot()
  });

  it('shows up on visible=true', () => {
    const wrapper = mount(<Toaster theme={themeMock} visible={true}/>)

    expect(wrapper.root()).toHaveStyleRule('opacity', '1')
    expect(wrapper.root()).toHaveStyleRule('z-index', themeMock.zIndexRank.top)
  });

  it('is hidden on visible=false', () => {
    const wrapper = mount(<Toaster theme={themeMock} visible={false}/>)

    expect(wrapper.root()).toHaveStyleRule('opacity', '0')
    expect(wrapper.root()).toHaveStyleRule('z-index', themeMock.zIndexRank.bottom)
  });

  it('renders provided text', () => {
    const wrapper = mount(
      <Toaster theme={themeMock} visible={false}>
        text
      </Toaster>
    )

    expect(wrapper.text()).toEqual('text')
  });
});
