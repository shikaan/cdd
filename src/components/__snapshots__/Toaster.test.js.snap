// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster renders correctly 1`] = `
<Portal
  containerInfo={<body />}
>
  <styled.div>
    <Component />
  </styled.div>
</Portal>
`;
